### ISSUE TITLE
feat(Seq): Add sequenceResultA

### ISSUE THREAD

**Issue by bartelink** [demystifyfp/FsToolkit.ErrorHandling#254]
There's a List.sequenceResultA

I'd find use for a `Seq.sequenceResultA: Result<'t, 'e> seq -> Result<'t[], 'e[]>`

This would align with how Async.Parallel works (accumulate outputs in a ResizeArray, yield an array) for efficiency

Example desired consumption

```
let res: Result<EmailAddress[], string> =
    seq { for i in Seq.distinct inputAddrs -> EmailAddress.tryParse i |> Result.requireSome i }
    |> Seq.sequenceResultA
    |> Result.mapError (fun es -> $"""Invalid inputs: {es |> String.concat ","}""" )
```

Example current workaround:

```
let res: Result<EmailAddress list, string> =
    [ for i in Seq.distinct inputAddrs -> EmailAddress.tryParse i |> Result.requireSome i ]
    |> List.sequenceResultA
    |> Result.mapError (fun es -> $"""Invalid inputs: {es |> String.concat ","}""" )
```

(This is literally what I have, and I have limited awareness of the Validation modules etc so maybe they offer a better workaround, but it does seem that [in general there's a hole in the completeness of the operations](https://stackoverflow.com/questions/50789065/turn-list-of-result-into-result-of-list-inside-a-computation-expression#comment137501687_71081606) hence raising this)

**Comment by TheAngryByrd** [comment]
Yep I agree this is a good idea. Would you like to submit a pull request for it?

**Comment by TheAngryByrd** [comment]
> I find the semantics of the laziness of that hard to reason abou

Yeah it seems like it's it's a Seq.fold, so I don't think it's going to be something that can lazily be evaluated. 

> Seq one could be obsoleted

Not necessarily because there are other enumerables beyond F#'s Array/List in the BCL and it's handy for those.

**Comment by bartelink** [comment]
Updates from spiking:
1. Fantomas default settings drive me wild (one pipe forcing a newline and other such nonsense). It's a real turnoff (and paket is IMO a bad idea for libraries). The ones TaskSeq are pretty tolerable. [The ones in FSharp.AWS.DyamoDB, which are a fork with 3 or 4 tweaks from TaskSeq, are very good](https://github.com/fsprojects/FSharp.AWS.DynamoDB/pull/73)
2. Do signatures need to be binary compatible? I'm going with putting it in the `Seq` module, and was going to make the `sequenceResultM` return an `array` on the basis that that's often useful, and it's what I intend to do for the `A` variant
3. I'm thinking not to do an `Array` module, as realistically one would want to have most of the `List` stuff in there too (but if no signature changes can be tolerated, then that's a way of sidestepping that issue). Also I am interested in having it the impl be tolerant of `seq` inputs, which does not totally align with that one would expect from an Array module.

**Comment by TheAngryByrd** [comment]
> one pipe forcing a newline and other such nonsense

I prefer this so this won't change.

> Do signatures need to be binary compatible? 

Yes unless you want to retarget for v5.

**Comment by bartelink** [comment]
>I prefer this so this won't change.

Yeah, ultimately defaults matter so not a problem (and I do get that lots of people are happy with that layout)

>Yes unless you want to retarget for v5.

I can make it API compatible in the context of this PR and prepare a follow-up PR to change the contract from `seq` to `[]` in a minimal way to be merged in the V5 timeframe

Also happy to leave the original impl exactly as it is; I have no doubt it's much closer to the idioms used in other impl pieces

I'll assume the former

**Comment by bartelink** [comment]
Sorry for dropping the ball on this - if you'd take the signature change, I'll do a PR to make V5 yield `Result<'t[], 'e[]>`.
That would allow removal of some Seq.toArray stuff and/or the use of Array.length and /or _.Length instead of Seq.length

aside: Did a local upgrade from 4.18 to V5 went in cleanly except for migrating some Async.retn to Async.singleton (good riddance to a bad name)

**Comment by TheAngryByrd** [comment]
I think there were two PRs around that code section and I probably got the merge wrong.

**Comment by bartelink** [comment]
OK, let me know if you want me to do anything (e.g. I could make a clone of the PR that was originally merged to v5 against some other branch of your choosing)

**Comment by TheAngryByrd** [comment]
Yes please do!
